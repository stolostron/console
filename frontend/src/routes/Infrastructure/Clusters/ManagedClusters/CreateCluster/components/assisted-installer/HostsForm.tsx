/* Copyright Contributors to the Open Cluster Management project */
import { useCallback, useRef, useEffect, useState, useMemo } from 'react'
import { CIM } from 'openshift-assisted-ui-lib'
import { useRecoilValue, waitForAll } from 'recoil'
import { FormikProps } from 'formik'
import { debounce, isEmpty, isEqual } from 'lodash'
import {
    agentClusterInstallsState,
    agentsState,
    clusterDeploymentsState,
    configMapsState,
} from '../../../../../../../atoms'
import { getAIConfigMap, onHostsNext } from './utils'

import './hosts-form.css'

const { ACMClusterDeploymentHostsStep, getTotalCompute, LoadingState } = CIM

type FormControl = {
    active?: CIM.ClusterDeploymentHostsSelectionValues
    agentClusterInstall: CIM.AgentClusterInstallK8sResource
    validate?: VoidFunction
    summary?: VoidFunction
    resourceJSON?: any
    step?: any
}

type HostsFormProps = {
    control: FormControl
    resourceJSON: any
    handleChange: (control: FormControl) => void
}

const HostsForm: React.FC<HostsFormProps> = ({ control, handleChange }) => {
    const [error, setError] = useState<string>()
    const formRef = useRef<FormikProps<CIM.ClusterDeploymentHostsSelectionValues>>(null)
    const [agents, clusterDeployments, agentClusterInstalls, configMaps] = useRecoilValue(
        waitForAll([agentsState, clusterDeploymentsState, agentClusterInstallsState, configMapsState])
    )
    const aiConfigMap = getAIConfigMap(configMaps)
    const { resourceJSON = {} } = control
    const { createResources = [] } = resourceJSON
    const cdName = createResources.find((r: { kind: string }) => r.kind === 'ClusterDeployment').metadata.name
    const aciName = createResources.find((r: { kind: string }) => r.kind === 'AgentClusterInstall').metadata.name

    const clusterDeployment = useMemo(
        () => clusterDeployments.find((cd) => cd.metadata.name === cdName && cd.metadata.namespace === cdName),
        [cdName, clusterDeployments]
    )
    const agentClusterInstall = useMemo(
        () => agentClusterInstalls.find((aci) => aci.metadata.name === aciName && aci.metadata.namespace === aciName),
        [aciName, agentClusterInstalls]
    )

    useEffect(() => {
        if (control.active && formRef?.current?.values && !isEqual(control.active, formRef.current.values)) {
            formRef?.current?.setValues(control.active, false)
        }
        control.validate = async () => {
            setError(undefined)
            formRef?.current?.setFieldError('patchError', undefined)
            const autoSelectHosts = formRef?.current?.values.autoSelectHosts
            const hostCount = formRef?.current?.values.hostCount
            const ids = formRef?.current?.values.autoSelectHosts
                ? formRef?.current?.values.autoSelectedHostIds
                : formRef?.current?.values.selectedHostIds
            const selectedAgents = agents.filter((a) => ids?.includes(a.metadata.uid))
            control.summary = () => {
                return [
                    {
                        term: 'Auto select hosts',
                        desc: autoSelectHosts ? 'Yes' : 'No',
                    },
                    {
                        term: 'Number of hosts',
                        desc: hostCount,
                    },
                    {
                        term: 'Total compute',
                        desc: getTotalCompute(selectedAgents),
                    },
                ]
            }
            await formRef?.current?.submitForm()
            if (!isEmpty(formRef?.current?.errors)) {
                return formRef?.current?.errors
            }

            try {
                await onHostsNext({ values: control.active, clusterDeployment, agents })
            } catch (err) {
                const msg = err instanceof Error ? err?.message : undefined
                setError(msg || 'An error occured')
                return {
                    resourcesError: 'Failed patching resources',
                }
            }
        }
    }, [control.active, clusterDeployment, agents])

    const onValuesChanged = useCallback(
        debounce((values) => {
            if (!isEqual(values, control.active)) {
                control.active = values
                control.step.title.isComplete = false
                handleChange(control)
            }
            // eslint-disable-next-line
        }, 300),
        []
    )

    return agents?.length && clusterDeployment && agentClusterInstall && aiConfigMap ? (
        <div className="hosts-form">
            <ACMClusterDeploymentHostsStep
                formRef={formRef}
                onValuesChanged={onValuesChanged}
                clusterDeployment={clusterDeployment}
                agentClusterInstall={agentClusterInstall}
                agents={agents}
                error={error}
                aiConfigMap={aiConfigMap}
            />
        </div>
    ) : (
        <LoadingState />
    )
}

export default HostsForm
