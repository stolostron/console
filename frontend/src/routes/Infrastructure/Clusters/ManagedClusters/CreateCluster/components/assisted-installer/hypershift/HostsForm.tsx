/* Copyright Contributors to the Open Cluster Management project */
import * as React from 'react'
import { HostedClusterHostsStep, LoadingState } from '@openshift-assisted/ui-lib/cim'
import { FormikProps } from 'formik'
import { HypershiftAgentContext } from './HypershiftAgentContext'
import { getClusterImageSet } from './utils'
import { useSharedAtoms, useSharedRecoil, useRecoilValue } from '../../../../../../../../shared-recoil'
import { useTranslation } from '../../../../../../../../lib/acm-i18next'
import { TFunction } from 'i18next'

type FormControl = {
  active: any // CIM.HostsFormValues
  disabled?: VoidFunction
  reverse?: (control: { active: any /* CIM.HostsFormValues */ }, templateObject: any) => void
  validate?: VoidFunction
  summary?: VoidFunction
  step?: any
  additionalProps?: { [x: string]: string }
  onNext: () => Promise<void>
}

type HostsFormProps = {
  control: FormControl
  handleChange: (control: FormControl) => void
}

export const getControlSummary = (activeControl: FormControl['active'], t: TFunction) => () =>
  [
    {
      term: t('Hosts namespace'),
      desc: activeControl.agentNamespace,
    },
    {
      term: t('Controller availability policy'),
      desc:
        activeControl.controllerAvailabilityPolicy === 'HighlyAvailable' ? t('Highly available') : t('Single replica'),
    },
    {
      term: t('Infrastructure availability policy'),
      desc:
        activeControl.infrastructureAvailabilityPolicy === 'HighlyAvailable'
          ? t('Highly available')
          : t('Single replica'),
    },
    {
      term: t('OLM catalog placement'),
      desc: activeControl.olmCatalogPlacement === 'management' ? t('Management') : t('Guest'),
    },
    {
      term: t('Node pools'),
      desc: activeControl.nodePools.length,
    },
    {
      term: t('Hosts count'),
      desc: activeControl.nodePools.reduce((acc: number, nodePool: any) => {
        acc += nodePool.useAutoscaling ? nodePool.autoscaling.maxReplicas : nodePool.count
        return acc
      }, 0),
    },
  ]

const HostsForm: React.FC<HostsFormProps> = ({ control, handleChange }) => {
  const {
    nodePools,
    setNodePools,
    clusterName,
    releaseImage,
    infraEnvNamespace,
    setInfraEnvNamespace,
    controllerAvailabilityPolicy,
    setControllerAvailabilityPolicy,
    infrastructureAvailabilityPolicy,
    setInfrastructureAvailabilityPolicy,
    olmCatalogPlacement,
    setOlmCatalogPlacement,
  } = React.useContext(HypershiftAgentContext)
  const { agentsState, clusterImageSetsState, infraEnvironmentsState, nodePoolsState } = useSharedAtoms()
  const { waitForAll } = useSharedRecoil()
  const [agents, infraEnvironments, clusterImageSets, currentNodePools] = useRecoilValue(
    waitForAll([agentsState, infraEnvironmentsState, clusterImageSetsState, nodePoolsState])
  )

  const formRef = React.useRef<FormikProps<any>>(null)
  const { t } = useTranslation()

  const initReleaseImage = getClusterImageSet(clusterImageSets, releaseImage)?.spec?.releaseImage

  const onValuesChanged = React.useCallback((values: any) => {
    if (Object.keys(values).some((key) => values[key] !== control.active?.[key])) {
      control.active = values
      control.step.title.isComplete = false
      setNodePools(values.nodePools)
      setInfraEnvNamespace(values.agentNamespace)
      setControllerAvailabilityPolicy(values.controllerAvailabilityPolicy)
      setInfrastructureAvailabilityPolicy(values.infrastructureAvailabilityPolicy)
      setOlmCatalogPlacement(values.olmCatalogPlacement)
      handleChange(control)
    }
    // eslint-disable-next-line
  }, [])

  control.validate = async () => {
    await formRef?.current?.submitForm()
    return formRef?.current?.errors
  }

  control.summary = getControlSummary(control.active, t)

  return agents ? (
    <HostedClusterHostsStep
      formRef={formRef}
      agents={agents}
      onValuesChanged={onValuesChanged}
      infraEnvs={infraEnvironments}
      clusterName={clusterName}
      initInfraEnv={infraEnvNamespace}
      initReleaseImage={initReleaseImage}
      initNodePools={nodePools}
      nodePools={currentNodePools}
      controllerAvailabilityPolicy={controllerAvailabilityPolicy}
      infrastructureAvailabilityPolicy={infrastructureAvailabilityPolicy}
      olmCatalogPlacement={olmCatalogPlacement}
    />
  ) : (
    <LoadingState />
  )
}

export default HostsForm
