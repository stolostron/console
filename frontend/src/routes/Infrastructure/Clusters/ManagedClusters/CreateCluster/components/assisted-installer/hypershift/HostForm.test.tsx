/* Copyright Contributors to the Open Cluster Management project */
import { MemoryRouter, Route } from 'react-router-dom'
import { RecoilRoot } from 'recoil'
import { NavigationPath } from '../../../../../../../../NavigationPath'
import HostsForm, { getControlSummary } from './HostsForm'
import { render } from '@testing-library/react'
import i18next from 'i18next'

describe('HostForm', () => {
  const handleChange = jest.fn()
  const onNext = jest.fn()
  const activeControl = {
    agentNamespace: 'test test',
    nodePools: [],
    controllerAvailabilityPolicy: 'HighlyAvailable',
    infrastructureAvailabilityPolicy: 'HighlyAvailable',
    olmCatalogPlacement: 'management',
  }

  const activeControl2 = {
    agentNamespace: 'test test',
    nodePools: [],
    controllerAvailabilityPolicy: 'SingleReplica',
    infrastructureAvailabilityPolicy: 'SingleReplica',
    olmCatalogPlacement: 'guest',
  }

  const t = i18next.t.bind(i18next)
  const Component = () => {
    return (
      <RecoilRoot>
        <MemoryRouter initialEntries={[NavigationPath.createCluster]}>
          <Route path={NavigationPath.createCluster}>
            <HostsForm
              control={{
                active: activeControl,
                step: {
                  title: {
                    isComplete: false,
                  },
                },
                onNext: onNext,
              }}
              handleChange={handleChange}
            />
          </Route>
        </MemoryRouter>
      </RecoilRoot>
    )
  }

  test('it renders', async () => {
    const { container } = render(<Component />)
    expect(container).toMatchSnapshot()
  })

  test('summary should match', async () => {
    const summary = getControlSummary(activeControl, t)
    expect(summary()).toMatchSnapshot()

    const summary2 = getControlSummary(activeControl2, t)
    expect(summary2()).toMatchSnapshot()
  })
})
