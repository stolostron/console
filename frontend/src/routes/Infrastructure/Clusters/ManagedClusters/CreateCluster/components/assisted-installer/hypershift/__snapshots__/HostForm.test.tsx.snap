// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostForm it renders 1`] = `
<div>
  <div
    class="pf-l-stack pf-m-gutter"
  >
    <div
      class="pf-l-stack__item"
    >
      ai:Define the quantity of worker nodes and nodepools to create for your cluster. Additional worker nodes and nodepools can be added after the cluster is created.
    </div>
    <div
      class="pf-l-stack__item"
    >
      <form
        class="pf-c-form"
        novalidate=""
      >
        <div
          class="pf-l-grid pf-m-gutter"
        >
          <div
            class="pf-l-grid__item"
          >
            <div
              class="pf-c-form__group"
            >
              <div
                class="pf-c-form__group-label"
              >
                <label
                  class="pf-c-form__label"
                >
                  <span
                    class="pf-c-form__label-text"
                  >
                    ai:Controller availability policy
                  </span>
                  <span
                    aria-hidden="true"
                    class="pf-c-form__label-required"
                  >
                     
                    *
                  </span>
                </label>
                 
              </div>
              <div
                class="pf-c-form__group-control pf-m-inline"
              >
                <div
                  class="pf-c-radio"
                >
                  <input
                    aria-invalid="false"
                    checked=""
                    class="pf-c-radio__input"
                    data-ouia-component-id="OUIA-Generated-Radio-1"
                    data-ouia-component-type="PF4/Radio"
                    data-ouia-safe="true"
                    id="form-radio-controllerAvailabilityPolicy-HighlyAvailable-field"
                    name="controllerAvailabilityPolicy"
                    type="radio"
                    value="HighlyAvailable"
                  />
                  <label
                    class="pf-c-radio__label"
                    for="form-radio-controllerAvailabilityPolicy-HighlyAvailable-field"
                  >
                    ai:Highly available
                     
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
                      data-ouia-component-id="OUIA-Generated-Button-plain-1"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                        />
                      </svg>
                    </button>
                  </label>
                </div>
                <div
                  class="pf-c-radio"
                >
                  <input
                    aria-invalid="false"
                    class="pf-c-radio__input"
                    data-ouia-component-id="OUIA-Generated-Radio-2"
                    data-ouia-component-type="PF4/Radio"
                    data-ouia-safe="true"
                    id="form-radio-controllerAvailabilityPolicy-SingleReplica-field"
                    name="controllerAvailabilityPolicy"
                    type="radio"
                    value="SingleReplica"
                  />
                  <label
                    class="pf-c-radio__label"
                    for="form-radio-controllerAvailabilityPolicy-SingleReplica-field"
                  >
                    ai:Single replica
                     
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
                      data-ouia-component-id="OUIA-Generated-Button-plain-2"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                        />
                      </svg>
                    </button>
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-l-grid__item"
          >
            <div
              class="pf-c-form__group"
            >
              <div
                class="pf-c-form__group-label"
              >
                <label
                  class="pf-c-form__label"
                >
                  <span
                    class="pf-c-form__label-text"
                  >
                    ai:Infrastructure availability policy
                  </span>
                  <span
                    aria-hidden="true"
                    class="pf-c-form__label-required"
                  >
                     
                    *
                  </span>
                </label>
                 
              </div>
              <div
                class="pf-c-form__group-control pf-m-inline"
              >
                <div
                  class="pf-c-radio"
                >
                  <input
                    aria-invalid="false"
                    checked=""
                    class="pf-c-radio__input"
                    data-ouia-component-id="OUIA-Generated-Radio-3"
                    data-ouia-component-type="PF4/Radio"
                    data-ouia-safe="true"
                    id="form-radio-infrastructureAvailabilityPolicy-HighlyAvailable-field"
                    name="infrastructureAvailabilityPolicy"
                    type="radio"
                    value="HighlyAvailable"
                  />
                  <label
                    class="pf-c-radio__label"
                    for="form-radio-infrastructureAvailabilityPolicy-HighlyAvailable-field"
                  >
                    ai:Highly available
                     
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
                      data-ouia-component-id="OUIA-Generated-Button-plain-3"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                        />
                      </svg>
                    </button>
                  </label>
                </div>
                <div
                  class="pf-c-radio"
                >
                  <input
                    aria-invalid="false"
                    class="pf-c-radio__input"
                    data-ouia-component-id="OUIA-Generated-Radio-4"
                    data-ouia-component-type="PF4/Radio"
                    data-ouia-safe="true"
                    id="form-radio-infrastructureAvailabilityPolicy-SingleReplica-field"
                    name="infrastructureAvailabilityPolicy"
                    type="radio"
                    value="SingleReplica"
                  />
                  <label
                    class="pf-c-radio__label"
                    for="form-radio-infrastructureAvailabilityPolicy-SingleReplica-field"
                  >
                    ai:Single replica
                     
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
                      data-ouia-component-id="OUIA-Generated-Button-plain-4"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                        />
                      </svg>
                    </button>
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-l-grid__item"
          >
            <div
              class="pf-c-form__group"
            >
              <div
                class="pf-c-form__group-label"
              >
                <label
                  class="pf-c-form__label"
                >
                  <span
                    class="pf-c-form__label-text"
                  >
                    ai:OLM catalog placement
                  </span>
                  <span
                    aria-hidden="true"
                    class="pf-c-form__label-required"
                  >
                     
                    *
                  </span>
                </label>
                 
              </div>
              <div
                class="pf-c-form__group-control pf-m-inline"
              >
                <div
                  class="pf-c-radio"
                >
                  <input
                    aria-invalid="false"
                    checked=""
                    class="pf-c-radio__input"
                    data-ouia-component-id="OUIA-Generated-Radio-5"
                    data-ouia-component-type="PF4/Radio"
                    data-ouia-safe="true"
                    id="form-radio-olmCatalogPlacement-management-field"
                    name="olmCatalogPlacement"
                    type="radio"
                    value="management"
                  />
                  <label
                    class="pf-c-radio__label"
                    for="form-radio-olmCatalogPlacement-management-field"
                  >
                    ai:Management
                     
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
                      data-ouia-component-id="OUIA-Generated-Button-plain-5"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                        />
                      </svg>
                    </button>
                  </label>
                </div>
                <div
                  class="pf-c-radio"
                >
                  <input
                    aria-invalid="false"
                    class="pf-c-radio__input"
                    data-ouia-component-id="OUIA-Generated-Radio-6"
                    data-ouia-component-type="PF4/Radio"
                    data-ouia-safe="true"
                    id="form-radio-olmCatalogPlacement-guest-field"
                    name="olmCatalogPlacement"
                    type="radio"
                    value="guest"
                  />
                  <label
                    class="pf-c-radio__label"
                    for="form-radio-olmCatalogPlacement-guest-field"
                  >
                    ai:Guest
                     
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
                      data-ouia-component-id="OUIA-Generated-Button-plain-6"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                        />
                      </svg>
                    </button>
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-l-grid__item"
          >
            <div
              class="pf-l-stack"
              id="form-control__form-input-agentNamespace-field"
            >
              <div
                class="pf-l-stack__item"
              >
                <div
                  class="pf-c-form__group"
                >
                  <div
                    class="pf-c-form__group-label"
                  >
                    <label
                      class="pf-c-form__label"
                      for="form-input-agentNamespace-field"
                    >
                      <span
                        class="pf-c-form__label-text"
                      >
                        ai:Namespace
                      </span>
                      <span
                        aria-hidden="true"
                        class="pf-c-form__label-required"
                      >
                         
                        *
                      </span>
                    </label>
                     
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-plain pf-c-form__group-label-help pf-u-p-0"
                      data-ouia-component-id="OUIA-Generated-Button-plain-7"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                        />
                      </svg>
                    </button>
                  </div>
                  <div
                    class="pf-c-form__group-control"
                  >
                    <select
                      aria-describedby="form-input-agentNamespace-field-helper"
                      aria-invalid="false"
                      class="pf-c-form-control"
                      data-ouia-component-id="OUIA-Generated-FormSelect-default-1"
                      data-ouia-component-type="PF4/FormSelect"
                      data-ouia-safe="true"
                      disabled=""
                      id="form-input-agentNamespace-field"
                      name="agentNamespace"
                      required=""
                    >
                      <option
                        class=""
                        value="NOT_AVAILABLE"
                      >
                        ai:No namespace with hosts is available
                      </option>
                    </select>
                    <div
                      aria-live="polite"
                      class="pf-c-form__helper-text"
                      id="form-input-agentNamespace-field-helper"
                    />
                  </div>
                </div>
              </div>
              <div
                class="pf-l-stack__item"
              />
            </div>
          </div>
          <div
            class="pf-l-grid__item"
          >
            <div
              aria-label="Warning Alert"
              class="pf-c-alert pf-m-inline pf-m-warning"
              data-ouia-component-id="OUIA-Generated-Alert-warning-1"
              data-ouia-component-type="PF4/Alert"
              data-ouia-safe="true"
            >
              <div
                class="pf-c-alert__icon"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 576 512"
                  width="1em"
                >
                  <path
                    d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                  />
                </svg>
              </div>
              <h4
                class="pf-c-alert__title"
              >
                <span
                  class="pf-u-screen-reader"
                >
                  Warning alert:
                </span>
                ai:The cluster has 0 hosts. No workloads will be able to run.
              </h4>
            </div>
          </div>
          <div
            class="pf-l-grid__item"
          >
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-link"
              data-ouia-component-id="OUIA-Generated-Button-link-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              ai:Add Nodepool
              <span
                class="pf-c-button__icon pf-m-end"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`HostForm summary should match 1`] = `
Array [
  Object {
    "desc": "test test",
    "term": "Hosts namespace",
  },
  Object {
    "desc": "Highly available",
    "term": "Controller availability policy",
  },
  Object {
    "desc": "Highly available",
    "term": "Infrastructure availability policy",
  },
  Object {
    "desc": "Management",
    "term": "OLM catalog placement",
  },
  Object {
    "desc": 0,
    "term": "Node pools",
  },
  Object {
    "desc": 0,
    "term": "Hosts count",
  },
]
`;

exports[`HostForm summary should match 2`] = `
Array [
  Object {
    "desc": "test test",
    "term": "Hosts namespace",
  },
  Object {
    "desc": "Single replica",
    "term": "Controller availability policy",
  },
  Object {
    "desc": "Single replica",
    "term": "Infrastructure availability policy",
  },
  Object {
    "desc": "Guest",
    "term": "OLM catalog placement",
  },
  Object {
    "desc": 0,
    "term": "Node pools",
  },
  Object {
    "desc": 0,
    "term": "Hosts count",
  },
]
`;
